jupyterlab_gcedetails/src/service/notebook_instance_service_layer.ts

```diff
@@ -0,0 +1,453 @@
+/**
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/* eslint-disable @typescript-eslint/camelcase */
+
+import {
+  handleApiError,
+  TransportService,
+  POST,
+  PATCH,
+  InstanceMetadata,
+  getMetadata,
+} from 'gcp_jupyterlab_shared';
+
+const POLL_INTERVAL = 5000;
+const POLL_RETRIES = 3;
+
+interface Service {
+  endpoint: string;
+  name: string;
+  documentation: string;
+  isOptional: boolean;
+}
+
+interface AcceleratorConfig {
+  type: string;
+  coreCount: string;
+}
+
+export interface Instance {
+  name: string;
+  state?: string;
+  machineType?: string;
+  acceleratorConfig?: AcceleratorConfig;
+}
+
+/** Service wrapper state corresponding to a Notebook Instance. */
+export interface State {
+  serviceEnabled: boolean;
+  projectId: string;
+  instanceName: string;
+  locationId: string;
+}
+
+/** API service status */
+export interface ServiceStatus {
+  service: Service;
+  enabled: boolean;
+}
+
+export const NOTEBOOKS_API_SERVICE: Readonly<Service> = {
+  name: 'AI Platform Notebooks API',
+  endpoint: 'notebooks.googleapis.com',
+  documentation: 'https://cloud.google.com/ai-platform/notebooks/',
+  isOptional: false,
+};
+
+type Operation = gapi.client.servicemanagement.Operation;
+type ListServicesResponse = gapi.client.servicemanagement.ListServicesResponse;
+
+const SERVICE_MANAGER = 'https://servicemanagement.googleapis.com/v1';
+export const NOTEBOOKS_API_PATH = 'https://notebooks.googleapis.com/v1beta1';
+
+const GOOGLE_API_UNAUTHENITCATED_CODE = 16;
+
+export function isUnauthorized(err: any) {
+  // Check for Google API Error structure and Unauthenticated error code
+  // https://cloud.google.com/apis/design/errors#error_codes
+  if (err.result && err.result.error) {
+    return err.result.error.code === GOOGLE_API_UNAUTHENITCATED_CODE;
+  }
+
+  return false;
+}
+
+/**
+ * Wraps around the GCP AI Platform Notebooks API to manage notebook resources in Google Cloud.
+ * https://cloud.google.com/ai-platform/notebooks/docs/reference/rest
+ */
+export class NotebookInstanceServiceLayer {
```

Comment:
I'd consider shortening this to something like NotebooksService and renaming the file accordingly.